File: types/java/AccountStatement.java
// AccountStatement.java

package io.quicktype;

/**
 * Statement for an account for a given billing cycle
 */
public class AccountStatement {
    private long accountID;
    private AccountOverview accountOverview;
    private AdditionalMinPayDetails additionalMinPayDetails;
    private BalanceSummary balanceSummary;
    private CycleSummary cycleSummary;
    private String effectiveAsOfDate;
    private MinPayDue minPayDue;
    private OpenToBuy openToBuy;
    private Payoff payoff;
    private long statementID;

    /**
     * The Canopy-generated ID for the account
     */
    public long getAccountID() { return accountID; }
    public void setAccountID(long value) { this.accountID = value; }

    public AccountOverview getAccountOverview() { return accountOverview; }
    public void setAccountOverview(AccountOverview value) { this.accountOverview = value; }

    public AdditionalMinPayDetails getAdditionalMinPayDetails() { return additionalMinPayDetails; }
    public void setAdditionalMinPayDetails(AdditionalMinPayDetails value) { this.additionalMinPayDetails = value; }

    public BalanceSummary getBalanceSummary() { return balanceSummary; }
    public void setBalanceSummary(BalanceSummary value) { this.balanceSummary = value; }

    public CycleSummary getCycleSummary() { return cycleSummary; }
    public void setCycleSummary(CycleSummary value) { this.cycleSummary = value; }

    /**
     * The `Date-Time` you would like the system to return the data as of. IE tell me what the
     * account information was as if I had asked on `2020-10-20 00:00:00 EST`. If empty it
     * defaults to current time.
     */
    public String getEffectiveAsOfDate() { return effectiveAsOfDate; }
    public void setEffectiveAsOfDate(String value) { this.effectiveAsOfDate = value; }

    public MinPayDue getMinPayDue() { return minPayDue; }
    public void setMinPayDue(MinPayDue value) { this.minPayDue = value; }

    public OpenToBuy getOpenToBuy() { return openToBuy; }
    public void setOpenToBuy(OpenToBuy value) { this.openToBuy = value; }

    public Payoff getPayoff() { return payoff; }
    public void setPayoff(Payoff value) { this.payoff = value; }

    /**
     * The Canopy-generated ID for the statement
     */
    public long getStatementID() { return statementID; }
    public void setStatementID(long value) { this.statementID = value; }
}

// AccountOverview.java

package io.quicktype;

public class AccountOverview {
    private String accountStatus;
    private String accountStatusSubtype;
    private String promoExclusiveEnd;
    private String promoInclusiveStart;
    private String promoPurchaseWindowExclusiveEnd;
    private String promoPurchaseWindowInclusiveStart;

    /**
     * The Status of the Account. Active upon account creation.
     */
    public String getAccountStatus() { return accountStatus; }
    public void setAccountStatus(String value) { this.accountStatus = value; }

    /**
     * The subtype of the Status of the Account. Null upon account creation.
     */
    public String getAccountStatusSubtype() { return accountStatusSubtype; }
    public void setAccountStatusSubtype(String value) { this.accountStatusSubtype = value; }

    /**
     * If applicable, the `Date-Time` the promo period for the account ends.
     */
    public String getPromoExclusiveEnd() { return promoExclusiveEnd; }
    public void setPromoExclusiveEnd(String value) { this.promoExclusiveEnd = value; }

    /**
     * If applicable, the `Date-Time` the promo period for the account starts.
     */
    public String getPromoInclusiveStart() { return promoInclusiveStart; }
    public void setPromoInclusiveStart(String value) { this.promoInclusiveStart = value; }

    /**
     * If applicable, the `Date-Time` the purchase window for the account ends.
     */
    public String getPromoPurchaseWindowExclusiveEnd() { return promoPurchaseWindowExclusiveEnd; }
    public void setPromoPurchaseWindowExclusiveEnd(String value) { this.promoPurchaseWindowExclusiveEnd = value; }

    /**
     * If applicable, the `Date-Time` the purchase window for the account ends.
     */
    public String getPromoPurchaseWindowInclusiveStart() { return promoPurchaseWindowInclusiveStart; }
    public void setPromoPurchaseWindowInclusiveStart(String value) { this.promoPurchaseWindowInclusiveStart = value; }
}

// AdditionalMinPayDetails.java

package io.quicktype;

public class AdditionalMinPayDetails {
    private Long minPayAmDeferredInterestCents;
    private Long minPayAmInterestCents;
    private long minPayChargesPrincipalCents;
    private long minPayDeferredCents;
    private long minPayFeesCents;
    private long minPayInterestCents;
    private long previousMinPayCents;

    /**
     * The current AM deferred interest balance of the line item. Canopy tracks deferred
     * interest during an amortization period separately from deferred interest accrued during a
     * revolving period.
     */
    public Long getMinPayAmDeferredInterestCents() { return minPayAmDeferredInterestCents; }
    public void setMinPayAmDeferredInterestCents(Long value) { this.minPayAmDeferredInterestCents = value; }

    /**
     * Total am interest due for the billing cycle.
     */
    public Long getMinPayAmInterestCents() { return minPayAmInterestCents; }
    public void setMinPayAmInterestCents(Long value) { this.minPayAmInterestCents = value; }

    /**
     * Total principal due for the billing cycle.
     */
    public long getMinPayChargesPrincipalCents() { return minPayChargesPrincipalCents; }
    public void setMinPayChargesPrincipalCents(long value) { this.minPayChargesPrincipalCents = value; }

    /**
     * Total deferred interest due for the billing cycle.
     */
    public long getMinPayDeferredCents() { return minPayDeferredCents; }
    public void setMinPayDeferredCents(long value) { this.minPayDeferredCents = value; }

    /**
     * Total fees due for the billing cycle.
     */
    public long getMinPayFeesCents() { return minPayFeesCents; }
    public void setMinPayFeesCents(long value) { this.minPayFeesCents = value; }

    /**
     * Total interest due for the billing cycle.
     */
    public long getMinPayInterestCents() { return minPayInterestCents; }
    public void setMinPayInterestCents(long value) { this.minPayInterestCents = value; }

    /**
     * Previous amounts due, including fees. This is a subset of min_pay_cents.
     */
    public long getPreviousMinPayCents() { return previousMinPayCents; }
    public void setPreviousMinPayCents(long value) { this.previousMinPayCents = value; }
}

// BalanceSummary.java

package io.quicktype;

public class BalanceSummary {
    private Long amDeferredInterestBalanceCents;
    private Long amInterestBalanceCents;
    private Long chargesPrincipalCents;
    private Long deferredInterestBalanceCents;
    private Long feesBalanceCents;
    private Long interestBalanceCents;
    private Long loansPrincipalCents;
    private Long totalBalanceCents;

    /**
     * The total deferred interest balance (in cents) associated with the account.
     */
    public Long getAmDeferredInterestBalanceCents() { return amDeferredInterestBalanceCents; }
    public void setAmDeferredInterestBalanceCents(Long value) { this.amDeferredInterestBalanceCents = value; }

    /**
     * Total AM interest balance for the account.
     */
    public Long getAmInterestBalanceCents() { return amInterestBalanceCents; }
    public void setAmInterestBalanceCents(Long value) { this.amInterestBalanceCents = value; }

    /**
     * Total principal balance for the account.
     */
    public Long getChargesPrincipalCents() { return chargesPrincipalCents; }
    public void setChargesPrincipalCents(Long value) { this.chargesPrincipalCents = value; }

    /**
     * Total deferred interest balance for the account.
     */
    public Long getDeferredInterestBalanceCents() { return deferredInterestBalanceCents; }
    public void setDeferredInterestBalanceCents(Long value) { this.deferredInterestBalanceCents = value; }

    /**
     * Total fee balance for the account.
     */
    public Long getFeesBalanceCents() { return feesBalanceCents; }
    public void setFeesBalanceCents(Long value) { this.feesBalanceCents = value; }

    /**
     * Total interest balance for the account.
     */
    public Long getInterestBalanceCents() { return interestBalanceCents; }
    public void setInterestBalanceCents(Long value) { this.interestBalanceCents = value; }

    /**
     * Total principal balance for loans for the account.
     */
    public Long getLoansPrincipalCents() { return loansPrincipalCents; }
    public void setLoansPrincipalCents(Long value) { this.loansPrincipalCents = value; }

    /**
     * The total balance (in cents) associated with the account.
     */
    public Long getTotalBalanceCents() { return totalBalanceCents; }
    public void setTotalBalanceCents(Long value) { this.totalBalanceCents = value; }
}

// CycleSummary.java

package io.quicktype;

public class CycleSummary {
    private Long cycleAmDeferredInterestCents;
    private Long cycleAmInterestCents;
    private Long cycleChargeReturnsCents;
    private Long cycleChargesCents;
    private Long cycleCreditAdjustmentsCents;
    private Long cycleDebitAdjustmentsCents;
    private Long cycleDeferredInterestCents;
    private String cycleExclusiveEnd;
    private String cycleInclusiveStart;
    private Long cycleInterestCents;
    private Long cycleLateFeesCents;
    private Long cycleLoansCents;
    private Long cyclePaymentReversalsCents;
    private Long cyclePaymentReversalsFeesCents;
    private Long cyclePaymentsCents;
    private Long cycleWaivedDeferredInterestCents;

    /**
     * Total interest accrued during the billing cycle. Canopy tracks deferred interest during
     * an amortization period separately from deferred interest accrued during a revolving
     * period.
     */
    public Long getCycleAmDeferredInterestCents() { return cycleAmDeferredInterestCents; }
    public void setCycleAmDeferredInterestCents(Long value) { this.cycleAmDeferredInterestCents = value; }

    /**
     * The current AM interest balance of the line item. Canopy tracks interest during an
     * amortization period separately from deferred interest accrued during a revolving period.
     */
    public Long getCycleAmInterestCents() { return cycleAmInterestCents; }
    public void setCycleAmInterestCents(Long value) { this.cycleAmInterestCents = value; }

    /**
     * Sum of all returns that occurred on the account during the billing cycle.
     */
    public Long getCycleChargeReturnsCents() { return cycleChargeReturnsCents; }
    public void setCycleChargeReturnsCents(Long value) { this.cycleChargeReturnsCents = value; }

    /**
     * Sum of all charges that occurred on the account during the billing cycle.
     */
    public Long getCycleChargesCents() { return cycleChargesCents; }
    public void setCycleChargesCents(Long value) { this.cycleChargesCents = value; }

    /**
     * Sum of all credit adjustment amounts that occurred on the account during the billing
     * cycle.
     */
    public Long getCycleCreditAdjustmentsCents() { return cycleCreditAdjustmentsCents; }
    public void setCycleCreditAdjustmentsCents(Long value) { this.cycleCreditAdjustmentsCents = value; }

    /**
     * Sum of all debit adjustment amounts that occurred on the account during the billing cycle.
     */
    public Long getCycleDebitAdjustmentsCents() { return cycleDebitAdjustmentsCents; }
    public void setCycleDebitAdjustmentsCents(Long value) { this.cycleDebitAdjustmentsCents = value; }

    /**
     * Total interest accrued during the billing cycle.
     */
    public Long getCycleDeferredInterestCents() { return cycleDeferredInterestCents; }
    public void setCycleDeferredInterestCents(Long value) { this.cycleDeferredInterestCents = value; }

    /**
     * The inclusive ending `Date-Time` that defines which transations are part of this
     * statement.
     */
    public String getCycleExclusiveEnd() { return cycleExclusiveEnd; }
    public void setCycleExclusiveEnd(String value) { this.cycleExclusiveEnd = value; }

    /**
     * The inclusive starting `Date-Time` that defines which transations are part of this
     * statement.
     */
    public String getCycleInclusiveStart() { return cycleInclusiveStart; }
    public void setCycleInclusiveStart(String value) { this.cycleInclusiveStart = value; }

    /**
     * Total interest accrued during the billing cycle.
     */
    public Long getCycleInterestCents() { return cycleInterestCents; }
    public void setCycleInterestCents(Long value) { this.cycleInterestCents = value; }

    /**
     * Total late fees incurred during the billing cycle.
     */
    public Long getCycleLateFeesCents() { return cycleLateFeesCents; }
    public void setCycleLateFeesCents(Long value) { this.cycleLateFeesCents = value; }

    /**
     * Sum of all loans amounts that were initiated on the account during the billing cycle.
     */
    public Long getCycleLoansCents() { return cycleLoansCents; }
    public void setCycleLoansCents(Long value) { this.cycleLoansCents = value; }

    /**
     * Sum of all payment reversal amounts that occurred on the account during the billing cycle.
     */
    public Long getCyclePaymentReversalsCents() { return cyclePaymentReversalsCents; }
    public void setCyclePaymentReversalsCents(Long value) { this.cyclePaymentReversalsCents = value; }

    /**
     * Total payment reversal fees incurred during the billing cycle.
     */
    public Long getCyclePaymentReversalsFeesCents() { return cyclePaymentReversalsFeesCents; }
    public void setCyclePaymentReversalsFeesCents(Long value) { this.cyclePaymentReversalsFeesCents = value; }

    /**
     * Sum of all payments that occurred on the account during the billing cycle.
     */
    public Long getCyclePaymentsCents() { return cyclePaymentsCents; }
    public void setCyclePaymentsCents(Long value) { this.cyclePaymentsCents = value; }

    /**
     * Total deferred interest that was forgiven on the account during the billing cycle.
     */
    public Long getCycleWaivedDeferredInterestCents() { return cycleWaivedDeferredInterestCents; }
    public void setCycleWaivedDeferredInterestCents(Long value) { this.cycleWaivedDeferredInterestCents = value; }
}

// MinPayDue.java

package io.quicktype;

public class MinPayDue {
    private long minPayCents;
    private String minPayDueAt;

    /**
     * Total amount due for the billing cycle, summing cycle principal, interest, deferred
     * interest, and fees outstanding.
     */
    public long getMinPayCents() { return minPayCents; }
    public void setMinPayCents(long value) { this.minPayCents = value; }

    /**
     * The `Date-Time` the payment for this billing cycle is due.
     */
    public String getMinPayDueAt() { return minPayDueAt; }
    public void setMinPayDueAt(String value) { this.minPayDueAt = value; }
}

// OpenToBuy.java

package io.quicktype;

public class OpenToBuy {
    private Long availableCreditCents;
    private Long creditLimitCents;
    private Long openToBuyCents;
    private Long totalChargesCents;

    /**
     * The total available credit balance (in cents) for the account.
     */
    public Long getAvailableCreditCents() { return availableCreditCents; }
    public void setAvailableCreditCents(Long value) { this.availableCreditCents = value; }

    /**
     * Total Amount (in cents) that this account can borrow.
     */
    public Long getCreditLimitCents() { return creditLimitCents; }
    public void setCreditLimitCents(Long value) { this.creditLimitCents = value; }

    /**
     * If applicable, the total amount of available funds for continued purchase following a
     * purchase window pattern, where payments made do not replenish amount available for
     * purchase.
     */
    public Long getOpenToBuyCents() { return openToBuyCents; }
    public void setOpenToBuyCents(Long value) { this.openToBuyCents = value; }

    /**
     * Sum of all charges that occurred on the account since account origination.
     */
    public Long getTotalChargesCents() { return totalChargesCents; }
    public void setTotalChargesCents(Long value) { this.totalChargesCents = value; }
}

// Payoff.java

package io.quicktype;

public class Payoff {
    private Long expectedRemainingPaymentAmountCents;
    private Long totalPayoffCents;

    /**
     * The total amount slated to pay off the account over the lifecycle of the loan.
     */
    public Long getExpectedRemainingPaymentAmountCents() { return expectedRemainingPaymentAmountCents; }
    public void setExpectedRemainingPaymentAmountCents(Long value) { this.expectedRemainingPaymentAmountCents = value; }

    /**
     * The total amount needed to pay off the loan at this exact moment.
     */
    public Long getTotalPayoffCents() { return totalPayoffCents; }
    public void setTotalPayoffCents(Long value) { this.totalPayoffCents = value; }
}

File: types/go/AccountStatement.go
// Statement for an account for a given billing cycle
type AccountStatement struct {
	AccountID               string                   `json:"account_id"`                          // The Canopy-generated ID for the account
	AccountOverview         *AccountOverview         `json:"account_overview,omitempty"`          
	AdditionalMinPayDetails *AdditionalMinPayDetails `json:"additional_min_pay_details,omitempty"`
	BalanceSummary          *BalanceSummary          `json:"balance_summary,omitempty"`           
	CycleSummary            *CycleSummary            `json:"cycle_summary,omitempty"`             
	EffectiveAsOfDate       string                   `json:"effective_as_of_date"`                // The `Date-Time` you would like the system to return the data as of. IE tell me what the; account information was as if I had asked on `2020-10-20 00:00:00 EST`. If empty it; defaults to current time.
	LineItems               []LineItem               `json:"line_items,omitempty"`                // A snapshot of all line items that occurred during the cycle and were `VALID` at the time; of Statement cut.
	MinPayDue               *MinPayDue               `json:"min_pay_due,omitempty"`               
	OpenToBuy               *OpenToBuy               `json:"open_to_buy,omitempty"`               
	Payoff                  *Payoff                  `json:"payoff,omitempty"`                    
	StatementID             string                   `json:"statement_id"`                        // The Canopy-generated ID for the statement
}

type AccountOverview struct {
	AccountStatus                     *string `json:"account_status,omitempty"`                       // The Status of the Account. Active upon account creation.
	AccountStatusSubtype              *string `json:"account_status_subtype,omitempty"`               // The subtype of the Status of the Account. Null upon account creation.
	PromoExclusiveEnd                 *string `json:"promo_exclusive_end,omitempty"`                  // If applicable, the `Date-Time` the promo period for the account ends.
	PromoInclusiveStart               *string `json:"promo_inclusive_start,omitempty"`                // If applicable, the `Date-Time` the promo period for the account starts.
	PromoPurchaseWindowExclusiveEnd   *string `json:"promo_purchase_window_exclusive_end,omitempty"`  // If applicable, the `Date-Time` the purchase window for the account ends.
	PromoPurchaseWindowInclusiveStart *string `json:"promo_purchase_window_inclusive_start,omitempty"`// If applicable, the `Date-Time` the purchase window for the account ends.
}

type AdditionalMinPayDetails struct {
	MinPayAmDeferredInterestCents *int64 `json:"min_pay_am_deferred_interest_cents,omitempty"`// The current AM deferred interest balance of the line item. Canopy tracks deferred; interest during an amortization period separately from deferred interest accrued during a; revolving period.
	MinPayAmInterestCents         *int64 `json:"min_pay_am_interest_cents,omitempty"`         // Total am interest due for the billing cycle.
	MinPayChargesPrincipalCents   int64  `json:"min_pay_charges_principal_cents"`             // Total principal due for the billing cycle.
	MinPayDeferredCents           int64  `json:"min_pay_deferred_cents"`                      // Total deferred interest due for the billing cycle.
	MinPayFeesCents               int64  `json:"min_pay_fees_cents"`                          // Total fees due for the billing cycle.
	MinPayInterestCents           int64  `json:"min_pay_interest_cents"`                      // Total interest due for the billing cycle.
	PreviousMinPayCents           int64  `json:"previous_min_pay_cents"`                      // Previous amounts due, including fees. This is a subset of min_pay_cents.
}

type BalanceSummary struct {
	AmDeferredInterestBalanceCents *int64 `json:"am_deferred_interest_balance_cents,omitempty"`// The total deferred interest balance (in cents) associated with the account.
	AmInterestBalanceCents         *int64 `json:"am_interest_balance_cents,omitempty"`         // Total AM interest balance for the account.
	ChargesPrincipalCents          *int64 `json:"charges_principal_cents,omitempty"`           // Total principal balance for the account.
	DeferredInterestBalanceCents   *int64 `json:"deferred_interest_balance_cents,omitempty"`   // Total deferred interest balance for the account.
	FeesBalanceCents               *int64 `json:"fees_balance_cents,omitempty"`                // Total fee balance for the account.
	InterestBalanceCents           *int64 `json:"interest_balance_cents,omitempty"`            // Total interest balance for the account.
	LoansPrincipalCents            *int64 `json:"loans_principal_cents,omitempty"`             // Total principal balance for loans for the account.
	TotalBalanceCents              *int64 `json:"total_balance_cents,omitempty"`               // The total balance (in cents) associated with the account.
}

type CycleSummary struct {
	CycleAmDeferredInterestCents     *int64  `json:"cycle_am_deferred_interest_cents,omitempty"`    // Total interest accrued during the billing cycle. Canopy tracks deferred interest during; an amortization period separately from deferred interest accrued during a revolving; period.
	CycleAmInterestCents             *int64  `json:"cycle_am_interest_cents,omitempty"`             // The current AM interest balance of the line item. Canopy tracks interest during an; amortization period separately from deferred interest accrued during a revolving period.
	CycleChargeReturnsCents          *int64  `json:"cycle_charge_returns_cents,omitempty"`          // Sum of all returns that occurred on the account during the billing cycle.
	CycleChargesCents                *int64  `json:"cycle_charges_cents,omitempty"`                 // Sum of all charges that occurred on the account during the billing cycle.
	CycleCreditAdjustmentsCents      *int64  `json:"cycle_credit_adjustments_cents,omitempty"`      // Sum of all credit adjustment amounts that occurred on the account during the billing; cycle.
	CycleDebitAdjustmentsCents       *int64  `json:"cycle_debit_adjustments_cents,omitempty"`       // Sum of all debit adjustment amounts that occurred on the account during the billing cycle.
	CycleDeferredInterestCents       *int64  `json:"cycle_deferred_interest_cents,omitempty"`       // Total interest accrued during the billing cycle.
	CycleExclusiveEnd                *string `json:"cycle_exclusive_end,omitempty"`                 // The inclusive ending `Date-Time` that defines which transations are part of this; statement.
	CycleInclusiveStart              *string `json:"cycle_inclusive_start,omitempty"`               // The inclusive starting `Date-Time` that defines which transations are part of this; statement.
	CycleInterestCents               *int64  `json:"cycle_interest_cents,omitempty"`                // Total interest accrued during the billing cycle.
	CycleLateFeesCents               *int64  `json:"cycle_late_fees_cents,omitempty"`               // Total late fees incurred during the billing cycle.
	CycleLoansCents                  *int64  `json:"cycle_loans_cents,omitempty"`                   // Sum of all loans amounts that were initiated on the account during the billing cycle.
	CyclePaymentReversalsCents       *int64  `json:"cycle_payment_reversals_cents,omitempty"`       // Sum of all payment reversal amounts that occurred on the account during the billing cycle.
	CyclePaymentReversalsFeesCents   *int64  `json:"cycle_payment_reversals_fees_cents,omitempty"`  // Total payment reversal fees incurred during the billing cycle.
	CyclePaymentsCents               *int64  `json:"cycle_payments_cents,omitempty"`                // Sum of all payments that occurred on the account during the billing cycle.
	CycleWaivedDeferredInterestCents *int64  `json:"cycle_waived_deferred_interest_cents,omitempty"`// Total deferred interest that was forgiven on the account during the billing cycle.
}

type LineItem struct {
	AccountID               int64                    `json:"account_id"`                         // The account associated with the line item
	CreatedAt               *string                  `json:"created_at,omitempty"`               // The `Date-Time` which the line item was created.
	EffectiveAt             *string                  `json:"effective_at,omitempty"`             // The `Date-Time` that this line item became/becomes active
	ExternalFields          []ExternalField          `json:"external_fields,omitempty"`          // An array of static references to fields in a third party system.
	IssuerProcessorMetadata *IssuerProcessorMetadata `json:"issuer_processor_metadata,omitempty"`
	LineItemID              int64                    `json:"line_item_id"`                       // The ID associated with the line item
	LineItemOverview        *LineItemOverview        `json:"line_item_overview,omitempty"`       
	LineItemSummary         *LineItemSummary         `json:"line_item_summary,omitempty"`        
	MerchantData            *MerchantData            `json:"merchant_data,omitempty"`            // Merchant information if applicable.
	ProductID               int64                    `json:"product_id"`                         // the Product ID of the account
}

type ExternalField struct {
	Key   *string `json:"key,omitempty"`  // Key - i.e. Name of the External Party
	Value *string `json:"value,omitempty"`// Value - i.e. External Account ID
}

type IssuerProcessorMetadata struct {
	Privacy *Privacy `json:"privacy,omitempty"`
}

type Privacy struct {
	LastFour *int64 `json:"last_four,omitempty"`// Last four digits of the card against which the spend was made.
}

type LineItemOverview struct {
	Description    *string         `json:"description,omitempty"`     // A description of this particular line item if any. More common for adjustments.
	LineItemStatus *LineItemStatus `json:"line_item_status,omitempty"`// the corresponding Status for a line item
	LineItemType   LineItemType    `json:"line_item_type"`            // The Line Item Type. i.e. `CHARGE`, `PAYMENT`.
}

type LineItemSummary struct {
	AmDeferredInterestBalanceCents *int64 `json:"am_deferred_interest_balance_cents,omitempty"`// The current AM deferred interest balance of the line item. Canopy tracks deferred; interest during an amortization period separately from deferred interest accrued during a; revolving period.
	AmInterestBalanceCents         *int64 `json:"am_interest_balance_cents,omitempty"`         // The current AM interest balance of the line item. Canopy tracks interest during an; amortization period separately from deferred interest accrued during a revolving period.
	BalanceCents                   int64  `json:"balance_cents"`                               // The current balance of the line item, which accounts for interest accrued per the; `product's` interest policy and the `account's` interest rate attribute.
	DeferredInterestBalanceCents   int64  `json:"deferred_interest_balance_cents"`             // he current deferred interest balance of the line item.
	InterestBalanceCents           int64  `json:"interest_balance_cents"`                      // The current interest balance of the line item.
	OriginalAmountCents            int64  `json:"original_amount_cents"`                       // The originating amount of money (in cents) relating to this line item.
	PrincipalCents                 int64  `json:"principal_cents"`                             // The principal balance of the line item.
	TotalInterestPaidToDateCents   int64  `json:"total_interest_paid_to_date_cents"`           // The sum (in cents) of all payments towards interest charges, if any, applied to this line; item to date
}

// Merchant information if applicable.
type MerchantData struct {
	ID          *string  `json:"id,omitempty"`          
	MccCode     *float64 `json:"mcc_code,omitempty"`    
	Name        *string  `json:"name,omitempty"`        
	PhoneNumber *string  `json:"phone_number,omitempty"`
}

type MinPayDue struct {
	MinPayCents int64  `json:"min_pay_cents"` // Total amount due for the billing cycle, summing cycle principal, interest, deferred; interest, and fees outstanding.
	MinPayDueAt string `json:"min_pay_due_at"`// The `Date-Time` the payment for this billing cycle is due.
}

type OpenToBuy struct {
	AvailableCreditCents *int64 `json:"available_credit_cents,omitempty"`// The total available credit balance (in cents) for the account.
	CreditLimitCents     *int64 `json:"credit_limit_cents,omitempty"`    // Total Amount (in cents) that this account can borrow.
	OpenToBuyCents       *int64 `json:"open_to_buy_cents,omitempty"`     // If applicable, the total amount of available funds for continued purchase following a; purchase window pattern, where payments made do not replenish amount available for; purchase.
	TotalChargesCents    *int64 `json:"total_charges_cents,omitempty"`   // Sum of all charges that occurred on the account since account origination.
}

type Payoff struct {
	ExpectedRemainingPaymentAmountCents *int64 `json:"expected_remaining_payment_amount_cents,omitempty"`// The total amount slated to pay off the account over the lifecycle of the loan.
	TotalPayoffCents                    *int64 `json:"total_payoff_cents,omitempty"`                     // The total amount needed to pay off the loan at this exact moment.
}

// the corresponding Status for a line item
type LineItemStatus string
const (
	Authorized LineItemStatus = "AUTHORIZED"
	Invalid LineItemStatus = "INVALID"
	Offset LineItemStatus = "OFFSET"
	Pending LineItemStatus = "PENDING"
	Posted LineItemStatus = "POSTED"
	Reversed LineItemStatus = "REVERSED"
	Rolled LineItemStatus = "ROLLED"
	SplitInvalid LineItemStatus = "SPLIT_INVALID"
	SplitValid LineItemStatus = "SPLIT_VALID"
	Valid LineItemStatus = "VALID"
	Void LineItemStatus = "VOID"
)

// The Line Item Type. i.e. `CHARGE`, `PAYMENT`.
type LineItemType string
const (
	Charge LineItemType = "CHARGE"
	CreditOffset LineItemType = "CREDIT_OFFSET"
	DebitOffset LineItemType = "DEBIT_OFFSET"
	DeferredInterest LineItemType = "DEFERRED_INTEREST"
	Fee LineItemType = "FEE"
	Interest LineItemType = "INTEREST"
	LateFee LineItemType = "LATE_FEE"
	Loan LineItemType = "LOAN"
	MinDue LineItemType = "MIN_DUE"
	MonthFee LineItemType = "MONTH_FEE"
	Payment LineItemType = "PAYMENT"
	PaymentSplit LineItemType = "PAYMENT_SPLIT"
	ProductInterest LineItemType = "PRODUCT_INTEREST"
	PromoEnd LineItemType = "PROMO_END"
	PurchaseWindowEnd LineItemType = "PURCHASE_WINDOW_END"
	ReturnCheckFee LineItemType = "RETURN_CHECK_FEE"
	Statement LineItemType = "STATEMENT"
	YearFee LineItemType = "YEAR_FEE"
)
